File: 1_fibonacci_number/fibonacci.py
"""
Author: Jehadur Rahman
Organization: CyArm
GitHub: https://github.com/JehadurRE
Website: jehadurre.me
Date: 2025-07-10 09:43 (Thu)
"""

def fibonacci_naive(n):
    if n <= 1:
        return n
    return fibonacci_naive(n - 1) + fibonacci_naive(n - 2)

def fibonacci_fast(n):
    if n <= 1:
        return n
    
    arr = [0] * (n + 1)
    arr[0] = 0
    arr[1] = 1
    
    for i in range(2, n + 1):
        arr[i] = arr[i - 1] + arr[i - 2]
    
    return arr[n]

def test_solution():
    assert fibonacci_fast(3) == 2
    assert fibonacci_fast(10) == 55
    for n in range(20):
        assert fibonacci_fast(n) == fibonacci_naive(n)

if __name__ == '__main__':
    input_n = int(input())
    # print(fibonacci_naive(input_n))  # Slow version
    # test_solution()  # Test the implementation
    print(fibonacci_fast(input_n))

File: 1_fibonacci_number/fibonacci.rb
#!/usr/bin/env ruby
# Author: Jehadur Rahman
# Organization: CyArm
# GitHub: https://github.com/JehadurRE
# Website: jehadurre.me
# Date: 2025-07-10 09:43 (Thu)

# by Andronik Ordian

def calc_fib_naive(n)
  return n if n <= 1
  calc_fib_naive(n - 1) + calc_fib_naive(n - 2)
end

def calc_fib(n)
  return n if n <= 1
  
  arr = Array.new(n + 1)
  arr[0] = 0
  arr[1] = 1
  
  (2..n).each do |i|
    arr[i] = arr[i - 1] + arr[i - 2]
  end
  
  arr[n]
end

def test_solution
  raise "Test failed" unless calc_fib(3) == 2
  raise "Test failed" unless calc_fib(10) == 55
  (0...20).each do |n|
    raise "Test failed for n=#{n}" unless calc_fib(n) == calc_fib_naive(n)
  end
end

if __FILE__ == $0
  n = gets.to_i
  # puts "#{calc_fib_naive(n)}"  # Slow version
  # test_solution  # Test the implementation
  puts "#{calc_fib(n)}"
end
